File: examples/xor/xor.py
import sys
import dynet as dy

#xsent = True
xsent = False

HIDDEN_SIZE = 8
ITERATIONS = 2000

m = dy.Model()
trainer = dy.SimpleSGDTrainer(m)

W = m.add_parameters((HIDDEN_SIZE, 2))
b = m.add_parameters(HIDDEN_SIZE)
V = m.add_parameters((1, HIDDEN_SIZE))
a = m.add_parameters(1)

if len(sys.argv) == 2:
  m.populate_from_textfile(sys.argv[1])

x = dy.vecInput(2)
y = dy.scalarInput(0)
h = dy.tanh((W*x) + b)
if xsent:
    y_pred = dy.logistic((V*h) + a)
    loss = dy.binary_log_loss(y_pred, y)
    T = 1
    F = 0
else:
    y_pred = (V*h) + a
    loss = dy.squared_distance(y_pred, y)
    T = 1
    F = -1


for iter in range(ITERATIONS):
    mloss = 0.0
    for mi in range(4):
        x1 = mi % 2
        x2 = (mi // 2) % 2
        x.set([T if x1 else F, T if x2 else F])
        y.set(T if x1 != x2 else F)
        mloss += loss.scalar_value()
        loss.backward()
        trainer.update()
    mloss /= 4.
    print("loss: %0.9f" % mloss)

x.set([F,T])
z = -(-y_pred)
print(z.scalar_value())

m.save("xor.pymodel")

dy.renew_cg()

x = dy.vecInput(2)
y = dy.scalarInput(0)
h = dy.tanh((W*x) + b)
if xsent:
    y_pred = dy.logistic((V*h) + a)
else:
    y_pred = (V*h) + a
x.set([T,F])
print("TF",y_pred.scalar_value())
x.set([F,F])
print("FF",y_pred.scalar_value())
x.set([T,T])
print("TT",y_pred.scalar_value())
x.set([F,T])
print("FT",y_pred.scalar_value())

File: contrib/rust/examples/xor/xor.rs
extern crate dynet;

use std::env;

use dynet::*;

fn main() {
    let args: Vec<String> = env::args().collect();
    dynet::initialize(&mut DynetParams::from_args(false));

    let ITERATIONS = 30;

    let mut m = ParameterCollection::new();
    let mut trainer = SimpleSGDTrainer::default(&mut m);

    let HIDDEN_SIZE = 8;
    let initializer = ParameterInitGlorot::default();
    let mut p_W = m.add_parameters([HIDDEN_SIZE, 2], &initializer);
    let mut p_b = m.add_parameters([HIDDEN_SIZE], &initializer);
    let mut p_V = m.add_parameters([1, HIDDEN_SIZE], &initializer);
    let mut p_a = m.add_parameters([1], &initializer);
    if args.len() == 2 {
        m.load(&args[1]);
    }

    let mut cg = ComputationGraph::new();
    for iter in 0..ITERATIONS {
        let mut loss = 0.;
        for mi in 0..4 {
            let x1 = mi % 2 == 1;
            let x2 = (mi / 2) % 2 == 1;
            let x_values = [if x1 { 1. } else { -1. }, if x2 { 1. } else { -1. }];
            let y_value = if x1 != x2 { 1. } else { -1. };

            cg.clear();
            let W = parameter(&mut cg, &mut p_W);
            let b = parameter(&mut cg, &mut p_b);
            let V = parameter(&mut cg, &mut p_V);
            let a = parameter(&mut cg, &mut p_a);
            let x = input(&mut cg, [2], &x_values);
            let y = input_scalar(&mut cg, y_value);
            let h = tanh(W * x + b);
            let y_pred = V * h + a;
            let loss_expr = squared_distance(y_pred, y);

            if iter == 0 && mi == 0 {
                cg.print_graphviz();
            }

            loss += cg.forward(&loss_expr).as_scalar();
            cg.backward(&loss_expr);
            trainer.update();
        }
        loss /= 4.;
        println!("E = {}", loss);
    }
    m.save("xor.model");
}
